sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/RemovableSymbolCollector.cs

```diff
@@ -0,0 +1,189 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Common;
+
+namespace SonarAnalyzer.Helpers
+{
+    internal class RemovableSymbolCollector : CSharpSyntaxWalker
```

Comment:
It would be neater if the "walker" part was implemented in an inner class that was only visible to the outer collector. The public API of the collector would then only be the public methods that intended to be called directly, and wouldn't include all public methods of CSharpSyntaxWalker.
It would also mean the logic in UnusedPrivateMembers.VisitDeclaringReferences method could be moved to the outer collector.